### ISSUE TITLE
Current scap-security-guide contains a bug which breaks use of key only accounts after 60 days

### ISSUE THREAD

**Issue by eseglem** [plus3it/ash-linux-formula#384]
### Problem Description:

It appears that `scap-security-guide` is installed without any version specified [here](https://github.com/plus3it/ash-linux-formula/blob/master/ash-linux/el7/VendorSTIG/packages.sls#L17). Currently that installs `0.1.66` which contains [this commit](https://github.com/rumch-se/ccc_content/commit/57043932ae44c733ae6b93c609d0ee5cbc9cc2d7). Basically it goes through `/etc/shadow` and sets a 60 day timeout on *every* account in it. Including ones without passwords. This means after 60 days accounts such as `maintuser` become unusable because they require changing a non existent password when you log in with your SSH keys. 

This likely affects anything that runs any version of `watchmaker`, since this will always install the latest version of the package. I have only been using CentOS 7, but looks like 8 would have the same issues if installing the same version of the package.

This does appear to be fixed in master and probably release `0.1.67`, however there is no package for it yet in the yum repo. Given that the repo skips from `0.1.63` to `0.1.66` which are ~7 months apart. I don't know if we will see a new release any time soon. 

Obviously, now that the issue is known it is possible for us to update install scripts to avoid it. However we only just discovered it, and have instances which are already affected. I know there are possible workarounds to get back into the affected system, though have not tested them yet. And any systems we haven't lost access to, we can log in and change the setting. I am more concerned with anyone else who likely isn't aware of this issue. I would hope something could be put in place to mitigate this more universally in the mean time. 

### Expected Behavior:

Accounts using only SSH keys for login should remain usable for 'break glass' type situations.

### Actual Behavior:

Any account using SSH keys to log in to the system becomes unusable after 60 days. 

### (Detailed) Steps to reproduce:

Launch an instance, run these formulas (via watchmaker in my case), wait 60 days and try to log in with your keys.

I assume you could disable ntp and change the date on the instance to see the same behavior, but I have not done that yet. 

You can also see that it will happen by viewing the `/etc/shadow` file without having to actually experience the issue. 

### (Optional) Fix recommendation:

Ideas:

- Pin to an older version of the package until we see a new release. I believe the previous version is: `scap-security-guide-0.1.63-1.el7.centos.noarch`
- Patch the xml file before running it. 
- Undo it after it runs. 
- Create an rpm for the newest version and include it in the extra repo so it is installed instead.

**Comment by lorengordon** [comment]
Hi @eseglem! Watchmaker doesn't actually use the content provided by `scap-security-guide`. At least, not by default. It uses the content distributed by [`scap-formula`](https://github.com/plus3it/scap-formula), which is currently pinned to what is provided by [v0.1.63](https://github.com/plus3it/scap-formula/blob/952727a9bbe831ef9207f7594702d0e898438e5c/.github/workflows/dependabot_hack.yml#L25) of the scap-security-guide... We have some [work-in-progress to update to v0.1.67](https://github.com/plus3it/scap-formula/pull/42), but it's a little stalled due to the switch to SCAP 1.3 and some issues we're having making the builds sufficiently reproducible...

So, I have to guess that you are doing something non-default in how you are setting up the watchmaker content?

**Comment by eseglem** [comment]
I am not very familiar with all of the pieces involved, so it has taken me quite a while to trace through everything yesterday and today. Based on your responses, I can see I was a bit off in my conclusions. I was not aware of [scap-formula](https://github.com/plus3it/scap-formula) or that it had its own xml. It looks like a change in that XML is the cause of our symptoms.

If there are any questions about how things are being ran, I do have a bunch of logs collected, but I no longer think they are relevant, so holding off on including them. 

This command is what is making the changes: `oscap xccdf eval --remediate --profile xccdf_org.ssgproject.content_profile_stig /root/scap/content/openscap/ssg-centos7-ds.xml`. I thought that file was coming from the `scap-security-guide` package. Which does get installed, but I guess is not used. It is likely coming from the other repo though.

The behavior in question appeared to have existed for only a short time in ComplianceAsCode/content. I now see that the behavior has changed multiple times recently. So, I am sure the version numbers were wrong. Sorry about that. The timeline is actually:

- [20221219](https://github.com/ComplianceAsCode/content/commit/7118d7d35c499ece08b6e3f3bda395bc61e8c93f) Changed `awk` to exclude accounts without passwords.
- [20230206](https://github.com/ComplianceAsCode/content/commit/7167e713bb26b7c153b8b4449a65fc8a72df4dde) Breaks that behavior.
- [20230330](https://github.com/ComplianceAsCode/content/commit/09582c02a01f84f841618a589f0c67dbe006ceba) Changes the `awk` to exclude accounts without passwords again.

The behavior contained in scap-formula's [ssg-centos7-ds.xml](https://github.com/plus3it/scap-formula/blob/master/scap/content/guides/openscap/ssg-centos7-ds.xml) no longer aligns with the current behavior in their master branch. scap-formula currently will change all accounts no matter what. I don't know exactly when this XML started being used, but this *seemed* to be a change within the last couple months, and appeared to line up with the above changes. So I incorrectly came to the conclusion it was using theirs.

This is the relevant part of the XML:

```
              <xccdf-1.2:Rule selected="false" id="xccdf_org.ssgproject.content_rule_accounts_password_set_max_life_existing" severity="medium">
                <xccdf-1.2:title>Set Existing Passwords Maximum Age</xccdf-1.2:title>
                <xccdf-1.2:description>Configure non-compliant accounts to enforce a <xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_var_accounts_maximum_age_login_defs" use="legacy"/>-day maximum password lifetime
restriction by running the following command:
<html:pre>$ sudo chage -M <xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_var_accounts_maximum_age_login_defs" use="legacy"/> <html:i>USER</html:i></html:pre></xccdf-1.2:description>
                <xccdf-1.2:reference href="https://public.cyber.mil/stigs/cci/">CCI-000199</xccdf-1.2:reference>
                <xccdf-1.2:reference href="http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r4.pdf">IA-5(f)</xccdf-1.2:reference>
                <xccdf-1.2:reference href="http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r4.pdf">IA-5(1)(d)</xccdf-1.2:reference>
                <xccdf-1.2:reference href="http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r4.pdf">CM-6(a)</xccdf-1.2:reference>
                <xccdf-1.2:reference href="https://public.cyber.mil/stigs/downloads/?_dl_facet_stigs=operating-systems%2Cgeneral-purpose-os">SRG-OS-000076-GPOS-00044</xccdf-1.2:reference>
                <xccdf-1.2:reference href="">SRG-OS-000076-VMM-000430</xccdf-1.2:reference>
                <xccdf-1.2:reference href="https://public.cyber.mil/stigs/srg-stig-tools/">SV-204421r603261_rule</xccdf-1.2:reference>
                <xccdf-1.2:rationale>Any password, no matter how complex, can eventually be cracked. Therefore,
passwords need to be changed periodically. If the operating system does
not limit the lifetime of passwords and force users to change their
passwords, there is the risk that the operating system passwords could be
compromised.</xccdf-1.2:rationale>
                <xccdf-1.2:fix system="urn:xccdf:fix:script:sh" id="accounts_password_set_max_life_existing" complexity="low" disruption="low" reboot="false" strategy="restrict">
var_accounts_maximum_age_login_defs='<xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_var_accounts_maximum_age_login_defs" use="legacy"/>'


while IFS= read -r i; do
    passwd -x $var_accounts_maximum_age_login_defs $i
done &lt;   &lt;(awk -v var="$var_accounts_maximum_age_login_defs" -F: '$5 &gt; var || $5 == "" {print $1}' /etc/shadow)
</xccdf-1.2:fix>
                <xccdf-1.2:check system="http://oval.mitre.org/XMLSchema/oval-definitions-5">
                  <xccdf-1.2:check-export export-name="oval:ssg-var_accounts_maximum_age_login_defs:var:1" value-id="xccdf_org.ssgproject.content_value_var_accounts_maximum_age_login_defs"/>
                  <xccdf-1.2:check-export export-name="oval:ssg-var_accounts_minimum_age_login_defs:var:1" value-id="xccdf_org.ssgproject.content_value_var_accounts_minimum_age_login_defs"/>
                  <xccdf-1.2:check-content-ref href="ssg-rhel7-oval.xml" name="oval:ssg-accounts_password_set_max_life_existing:def:1"/>
                </xccdf-1.2:check>
                <xccdf-1.2:check system="http://scap.nist.gov/schema/ocil/2">
                  <xccdf-1.2:check-content-ref href="ssg-rhel7-ocil.xml" name="ocil:ssg-accounts_password_set_max_life_existing_ocil:questionnaire:1"/>
                </xccdf-1.2:check>
              </xccdf-1.2:Rule>
```

Other rules ensure settings in various locations are set in case any new accounts are created with a password. So, the OS is enforcing it. And **if** any accounts were to have a password set, ages would need to be set on them. However, none of the existing accounts have passwords, so this rule **should not** be applicable to any of them. A non-existent password cannot be vulnerable or cracked, nor should it ever need to be 'changed' since it doesn't exist in the first place. And their latest behavior would align with that.

If trying to be extra careful, could do things like `usermod -L maintuser` to prevent it from setting a password and add the following to `sshd_config` so it cannot log in with a password:

```
Match user maintuser
PasswordAuthentication no
```

**Comment by lorengordon** [comment]
Thanks for the investigation @eseglem. We're still trying to get the scap-formula updated with the content from the latest release. Can you see if v0.1.67 has the desired behavior?

(The ComplianceAsCode/content project changed a few things in their build process, and our build script doesn't seem to be working consistently, so it's taking us some time to understand what is happening under the covers so we can get it working again.)

**Comment by eseglem** [comment]
I grabbed [scap-security-guide-0.1.67.zip](https://github.com/ComplianceAsCode/content/releases/download/v0.1.67/scap-security-guide-0.1.67.zip) and pulled out `ssg-centos7-ds.xml` and `ssg-centos7-ds-1.2.xml` they both appear to have the latest behavior in them. 

```
                <xccdf-1.2:fix system="urn:xccdf:fix:script:sh" id="accounts_password_set_max_life_existing" complexity="low" disruption="low" reboot="false" strategy="restrict">
var_accounts_maximum_age_login_defs='<xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_var_accounts_maximum_age_login_defs" use="legacy"/>'

while IFS= read -r i; do
    chage -M $var_accounts_maximum_age_login_defs $i
done &lt;   &lt;(awk -v var="$var_accounts_maximum_age_login_defs" -F: '(/^[^:]+:[^!*]/ &amp;&amp; ($5 &gt; var || $5 == "")) {print $1}' /etc/shadow)
</xccdf-1.2:fix>
```

I did a little testing and believe this would not have the same issues. I logged into a dev instance that had ran everything and did the following:

```
# cat /etc/shadow | grep maintuser
maintuser:!!:19472:1:60:7:::
# chage -M -1 -m -1 maintuser
# cat /etc/shadow | grep maintuser
maintuser:!!:19472:::7:::
# awk -v var=60 -F: '$5 > var || $5 == "" {print $1}' /etc/shadow
maintuser
# awk -v var=60 -F: '(/^[^:]+:[^!*]/ && ($5 > var || $5 == "")) {print $1}' /etc/shadow
# 
```

So, best I can tell the new `awk` does not care about the maintuser account and wouldn't run the `chage` command in the first place.

**Comment by ferricoxide** [comment]
Looking at @eseglem's final response, it looks like this becomes a non-issue in the [v0.1.67](https://github.com/ComplianceAsCode/content/releases/v0.1.67/) CAC update? If so, can we close this issue?

**Comment by lorengordon** [comment]
We're still working to get the scap-formula updated so it is distributing the v0.1.67 content. Made some progress this week, reworking how we build the content to use a Dockerfile and a script that builds the image and pulls the content out of the image. Expecting that to improve reproducibility of the built content. Need to get verification on that, push the new content to the pr, test it against watchmaker, and then prioritize a watchmaker release. Hoping to get that done sometime next week...?

**Comment by lorengordon** [comment]
@eseglem We just released Watchmaker 0.27.1, which is using the v0.1.67 content from the ComplianceAsCode project. Can you please test with that version and let us know if you still see the problem?

**Comment by eseglem** [comment]
Thank you for your responsiveness and all the work you do on these tools. It is greatly appreciated. 

I'm am fairly confident this was addressed by the update. An instance I deployed earlier has 99999 for the max age for all the built in accounts, vs 60 in the earlier release. I did have safeties in place around maintuser to set it back to 99999 if it changed, but based on the userdata and audit logs it doesn't look like they triggered. So, should all be good.
